tools/ReadConsoleInputStream/ReadConsoleInputStream.cs

```diff
@@ -0,0 +1,194 @@
+        {
+            get => throw new NotSupportedException("Seek not supported.");
+            set => throw new NotSupportedException("Seek not supported.");
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _handle = IntPtr.Zero;
+
+            base.Dispose(disposing);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            ValidateRead(buffer, offset, count);
+
+            Debug.Assert(offset >= 0, "offset >= 0");
+            Debug.Assert(count >= 0, "count >= 0");
+            Debug.Assert(buffer != null, "bytes != null");
+
+            // Don't corrupt memory when multiple threads are erroneously writing
+            // to this stream simultaneously.
+            if (buffer.Length - offset < count)
+                throw new IndexOutOfRangeException("IndexOutOfRange_IORaceCondition");
+
+            int bytesRead;
+            int ret;
+
+            if (buffer.Length == 0)
+            {
+                bytesRead = 0;
+                ret = Win32Error.ERROR_SUCCESS;
+            }
+            else
+            {
+                var charsRead = 0;
+                bytesRead = 0;
+
+                var records = new Kernel32.INPUT_RECORD[BufferSize];
+
+                // begin input loop
+                waitForInput:
+
+                var readSuccess = Kernel32.ReadConsoleInput(_handle, records, 256, out var recordsRead);
+                Debug.WriteLine("Read {0} input record(s)", recordsRead);
+
+                if (readSuccess && recordsRead > 0)
+                {
+                    for (var index = 0; index < recordsRead; index++)
+                    {
+                        var record = records[index];
+
+                        if (record.EventType == Kernel32.EVENT_TYPE.KEY_EVENT)
+                        {
+                            // skip key up events - if not, every key will be duped in the stream
+                            if (record.Event.KeyEvent.bKeyDown == false) continue;
+
+                            // pack ucs-2/utf-16le/unicode chars into position in our byte[] buffer.
+                            var glyph = (ushort) record.Event.KeyEvent.uChar;
+
+                            var lsb = (byte) (glyph & 0xFFu);
+                            var msb = (byte) ((glyph >> 8) & 0xFFu);
+
+                            // ensure we accommodate key repeat counts
+                            for (var n = 0; n < record.Event.KeyEvent.wRepeatCount; n++)
+                            {
+                                buffer[offset + charsRead * BytesPerWChar] = lsb;
+                                buffer[offset + charsRead * BytesPerWChar + 1] = msb;
+
+                                charsRead++;
+                            }
+                        }
+                        else
+                        {
+                            // ignore focus events (not doing so makes debugging absolutely hilarious)
```

Comment:
>absolutely hilarious) [](start = 81, length = 21)

I mean true... but explain? "makes debugging absolutely hilarious as another focus event will be dispatched every time your debugger takes focus for a breakpoint and then again when it releases it to reactivate your application..." or something

Reply:
How long did you have to think about my comment before figuring out what I meant? :)

Reply:
Not long, but I'm not a fair metric here. I've done this so many times to myself that the conhost.exe code has a `#ifdef DBG` around focus events to block them from being put into the input queue to avoid this problem when I debug it.